checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -0,0 +1,190 @@
+    // by the Object Construction Checker's -AcheckMustCall algorithm.
+    if (!checker.hasOption(MustCallChecker.NO_RESOURCE_ALIASES)) {
+      useType.removeAnnotationByClass(MustCallAlias.class);
+    }
+    return super.isValidUse(declarationType, useType, tree);
+  }
+
+  @Override
+  protected boolean skipReceiverSubtypeCheck(
+      MethodInvocationTree node,
+      AnnotatedTypeMirror methodDefinitionReceiver,
+      AnnotatedTypeMirror methodCallReceiver) {
+    // It does not make sense for receivers to have must-call obligations. If the receiver of a
+    // method were to have a non-empty must-call obligation, then actually the method **IS** a
+    // must-call method! So skipping this check is always sound.
+    return true;
+  }
+
+  /**
+   * Mark (using the extraArgs parameter) any assignments where the LHS is a resource variable, so
+   * that close doesn't need to be considered. See {@link
+   * #commonAssignmentCheck(AnnotatedTypeMirror, AnnotatedTypeMirror, Tree, String, Object...)} for
+   * the code that uses and removes the marks.
+   */
+  @Override
+  protected void commonAssignmentCheck(
+      Tree varTree,
+      ExpressionTree valueExp,
+      @CompilerMessageKey String errorKey,
+      Object... extraArgs) {
+    if (TreeUtils.elementFromTree(varTree).getKind() == ElementKind.RESOURCE_VARIABLE) {
+      // Use the extraArgs array to signal to later stages of the common assignment check that this
+      // is in a
+      // resource variable context.
+      Object[] newExtraArgs = Arrays.copyOf(extraArgs, extraArgs.length + 1);
+      newExtraArgs[newExtraArgs.length - 1] = ElementKind.RESOURCE_VARIABLE;
+      super.commonAssignmentCheck(varTree, valueExp, errorKey, newExtraArgs);
+    } else {
+      super.commonAssignmentCheck(varTree, valueExp, errorKey, extraArgs);
+    }
+  }
+
+  /**
+   * If the LHS has been marked as a resource variable, then the standard common assignment check is
+   * skipped and a check that does not include "close" is substituted. Resource variables are marked
+   * by {@link #commonAssignmentCheck(Tree, ExpressionTree, String, Object...)} using the extraArgs
+   * parameter.
+   */
+  @Override
+  protected void commonAssignmentCheck(
+      AnnotatedTypeMirror varType,
+      AnnotatedTypeMirror valueType,
+      Tree valueTree,
+      @CompilerMessageKey String errorKey,
+      Object... extraArgs) {
+    if (extraArgs.length >= 1 && extraArgs[extraArgs.length - 1] == ElementKind.RESOURCE_VARIABLE) {
+      AnnotationMirror varAnno = varType.getAnnotationInHierarchy(atypeFactory.TOP);
+      AnnotationMirror valAnno = valueType.getAnnotationInHierarchy(atypeFactory.TOP);
+      if (atypeFactory
+          .getQualifierHierarchy()
+          .isSubtype(atypeFactory.withoutClose(valAnno), atypeFactory.withoutClose(varAnno))) {
+        return;
+      } else {
+        // Remove the resource variable entry, since otherwise the rest of the framework could
```

Comment:
What is the reason for the fallthrough here?  Is the `super` call guaranteed to fail and you want it to do so?  If so, please document that, though it may be clearer to issue a warning here.

Reply:
>  the super call guaranteed to fail and you want it to do so

Yes, unless the rest of the common assignment check has some kind of exception for the current case. That's an advantage of having a single common assignment check sequence - I can insert this exception for resource variables without changing how the rest of the check is computed, or disrupting e.g. the logging for `-Ashowchecks`.

Issuing an error would be counterproductive, especially from a maintenance perspective: there's already code to issue errors in the super call. I will add a note, however.